### ISSUE TITLE
native libraries experiment

### ISSUE THREAD

**Issue by borkdude** [borkdude/babashka#397]
Maybe it's possible to load a native library in babashka?

Here are some Graal docs:

https://github.com/oracle/graal/blob/master/substratevm/ImplementingNativeMethodsInJavaWithSVM.md

Here is some code to look at for inspiration (for lack of docs): https://github.com/graalvm/graalvm-demos/blob/master/native-netty-plot/src/main/java/com/oracle/svm/nettyplot/FunctionPlotter.java#L81

Jump to https://github.com/borkdude/babashka/issues/397#issuecomment-621406120 to see how to run the code from the `modules` branch.

**Comment by borkdude** [comment]
![Screenshot 2020-04-29 16 43 07](https://user-images.githubusercontent.com/284934/80609927-fd423e00-8a38-11ea-807d-8656f1f84038.png)

**Comment by jeroenvandijk** [comment]
Maybe architectural wise [boot pods](https://github.com/boot-clj/boot/wiki/Pods) could serve as a good example for an API. Pods are also limited in what they can share, but you can do rpc like calls that can run inside the pod and return. They have a similar constraint:
```
(with-eval-in pod body*)
Evaluate the body expressions in the pod and return the result. The body expressions are templates.

The body and result must be able to be printed by pr-str and then read by read-string.
```

**Comment by borkdude** [comment]
New attempt:

- Clone this repo, recursively: 

```
$ git clone https://github.com/borkdude/babashka --recursive
```

- Go into the `libhsqldb` directory. Run `./compile`. You need GraalVM 19.3.1 Java 8. This will produce `libhsqldb.babashka.org.dylib`.
- Go to the project root and run `lein javac`.
- Go into a REPL:
```
$ clj -Sdeps '{:paths ["target/classes"]}' -J-Djava.library.path=$PWD/libhsqldb
Clojure 1.10.2-alpha1
user=> (clojure.lang.RT/loadLibrary "hsqldb.babaska.org")
nil
user=> (org.babashka.LibHsqlDb/eval "(+ 1 2 3)")
Execution error (UnsatisfiedLinkError) at org.babashka.LibHsqlDb/createIsolate (LibHsqlDb.java:-2).
org.babashka.LibHsqlDb.createIsolate()J
user=>
```

Why isn't the above working? The original example at least didn't have this unsatisfied link error. Did I make a typo?

**Comment by jeroenvandijk** [comment]
Does it need to be `public final class` (add `final`) maybe? ([source](https://github.com/oracle/graal/blob/master/substratevm/ImplementingNativeMethodsInJavaWithSVM.md#initializing-the-substrate-vm))

I see that also in your jdbc example:
```
public final class LibSci {
    @CEntryPoint(name = "Java_babashka_modules_PgNextJDBC_initPG")
```

[This](https://github.com/borkdude/babashka/compare/modules#diff-5eff730ea7eb6674d88e979fe668c26fR3-R5) should be made `final` then

**Comment by borkdude** [comment]
@jeroenvandijk Did not help. Returning a `String` from the native library to the JVM calling it: how hard can it be? I updated the source. Now I'm getting:

```
$ clj -Sdeps '{:paths ["target/classes"]}' -J-Djava.library.path=$PWD/libhsqldb
Clojure 1.10.2-alpha1
user=> (clojure.lang.RT/loadLibrary "hsqldb.babashka.org")
nil
user=> (org.babashka.LibHsqlDb/eval "(+ 1 2 3)")
x
y
debug: xxx
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000010e5959b5, pid=62093, tid=0x0000000000001a03
#
# JRE version: OpenJDK Runtime Environment (8.0_212-b04) (build 1.8.0_212-b04)
# Java VM: OpenJDK 64-Bit Server VM (25.212-b04 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# J 735 C2 java.lang.StringBuilder.append(Ljava/lang/String;)Ljava/lang/StringBuilder; (8 bytes) @ 0x000000010e5959b5 [0x000000010e595980+0x35]
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/borkdude/Dropbox/dev/clojure/babashka/hs_err_pid62093.log
Compiled method (c2)    4328  735       4       java.lang.StringBuilder::append (8 bytes)
 total in heap  [0x000000010e595810,0x000000010e5961b0] = 2464
 relocation     [0x000000010e595938,0x000000010e595970] = 56
 main code      [0x000000010e595980,0x000000010e595dc0] = 1088
 stub code      [0x000000010e595dc0,0x000000010e595dd8] = 24
 metadata       [0x000000010e595dd8,0x000000010e595e18] = 64
 scopes data    [0x000000010e595e18,0x000000010e596060] = 584
 scopes pcs     [0x000000010e596060,0x000000010e596130] = 208
 dependencies   [0x000000010e596130,0x000000010e596138] = 8
 handler table  [0x000000010e596138,0x000000010e596180] = 72
 nul chk table  [0x000000010e596180,0x000000010e5961b0] = 48
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
clojure: warning: clojure crashed, killed by SIGABRT.
clojure itself has not crashed, but for transparency,
it will now kill itself (without dumping core) with the same signal
```

**Comment by jeroenvandijk** [comment]
`evalString` which is called by `eval` doesn't return a `String` but a `org.graalvm.nativeimage.c.type.CCharPointer` ([source](https://github.com/borkdude/babashka/compare/modules#diff-fc7ad20dbdbf8acc8f4ea9c0b90868bcR23-R25))

@borkdude Could that be the issue? Maybe it's enough to convert the char array to a String. Given the error message `String.append.. something` this sounds as the potential problem

**Comment by borkdude** [comment]
@jeroenvandijk Yes, that's probably it. I'm now testing this:

```
package org.babashka;

import org.graalvm.nativeimage.c.type.CCharPointer;
import com.oracle.svm.core.c.CConst;
import org.graalvm.nativeimage.c.type.CTypeConversion;

public final class LibHsqlDb {
    private static native @CConst CCharPointer evalString(long isolateThreadId);
    private static native long createIsolate();
    private static native void tearDownIsolate(long id);

    public static void eval(String expr) {
        System.out.println("x");
        System.out.flush();
        long isolate = createIsolate();
        System.out.println("y");
        System.out.flush();
        CCharPointer c = evalString(isolate);
        String x = CTypeConversion.toJavaString(c);
        //System.out.println("x: " + x);
        System.out.flush();
        //tearDownIsolate(isolate);
        //return x;
    }
}
```

```
$ clj -Sdeps '{:deps {org.graalvm.nativeimage/svm {:mvn/version "19.3.1"}} :paths ["target/classes"]}' -J-Djava.library.path=$PWD/libhsqldb
Clojure 1.10.2-alpha1
user=> (clojure.lang.RT/loadLibrary "hsqldb.babashka.org")
nil
user=> (org.babashka.LibHsqlDb/eval "(+ 1 2 3)")
x
y
debug: xxx
Execution error (Error) at org.graalvm.nativeimage.impl.ImageSingletonsSupport/checkInstalled (ImageSingletonsSupport.java:69).
The class ImageSingletons can only be used when building native images, i.e., when using the native-image command.
```

**Comment by borkdude** [comment]
@jeroenvandijk I inlined that `toJavaString` method:

```
package org.babashka;

import org.graalvm.nativeimage.c.type.CCharPointer;
import com.oracle.svm.core.c.CConst;
import org.graalvm.nativeimage.c.type.CTypeConversion;
import org.graalvm.word.WordFactory;
import org.graalvm.word.UnsignedWord;
import org.graalvm.word.Pointer;

public final class LibHsqlDb {
    private static native @CConst CCharPointer evalString(long isolateThreadId);
    private static native long createIsolate();
    private static native void tearDownIsolate(long id);

    private static String toJavaStringUnchecked(CCharPointer cString, UnsignedWord length) {
        byte[] bytes = new byte[(int) length.rawValue()];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = ((Pointer) cString).readByte(i);
        }
        return new String(bytes);
    }

    public static UnsignedWord strlen(CCharPointer str) {
        UnsignedWord n = WordFactory.zero();
        while (((Pointer) str).readByte(n) != 0) {
            n = n.add(1);
        }
        return n;
    }

    public static String toJavaString(CCharPointer cString) {
        if (cString.isNull()) {
            return null;
        } else {
            return toJavaStringUnchecked(cString, strlen(cString));
        }
    }

    public static void eval(String expr) {
        System.out.println("x");
        System.out.flush();
        long isolate = createIsolate();
        System.out.println("y");
        System.out.flush();
        CCharPointer c = evalString(isolate);
        String x = toJavaString(c);
        //System.out.println("x: " + x);
        System.out.flush();
        //tearDownIsolate(isolate);
        //return x;
    }
}
```

```
$ clj -Sdeps '{:deps {org.graalvm.nativeimage/svm {:mvn/version "19.3.1"}} :paths ["target/classes"]}' -J-Djava.library.path=$PWD/libhsqldb
Clojure 1.10.2-alpha1
user=> (clojure.lang.RT/loadLibrary "hsqldb.babashka.org")
nil
user=> (org.babashka.LibHsqlDb/eval "(+ 1 2 3)")
x
y
debug: xxx
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001043a5a9a, pid=63929, tid=0x0000000000001c03
#
# JRE version: OpenJDK Runtime Environment (8.0_212-b04) (build 1.8.0_212-b04)
# Java VM: OpenJDK 64-Bit Server VM (25.212-b04 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.dylib+0x3a5a9a]  LinkResolver::resolve_invokeinterface(CallInfo&, Handle, constantPoolHandle, int, Thread*)+0x100
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/borkdude/Dropbox/dev/clojure/babashka/hs_err_pid63929.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
clojure: warning: clojure crashed, killed by SIGABRT.
clojure itself has not crashed, but for transparency,
it will now kill itself (without dumping core) with the same signal


warnings can be silenced by the --no-warnings (-n) option
[1]    63928 abort      clj -Sdeps  -J-Djava.library.path=$PWD/libhsqldb
```

**Comment by jeroenvandijk** [comment]
Feels like you are really close. Where does it crash? Are you able to print the last string? Maybe the `tearDownIsolate` needs to be activated again? ([source](https://github.com/graalvm/graalvm-demos/blob/master/native-netty-plot/src/main/java/com/oracle/svm/nettyplot/FunctionPlotter.java#L88))

**Comment by borkdude** [comment]
@jeroenvandijk Adding it back didn't help. In fact, I commented it out because I thought it could be  the cause of the error. It seems the error is a segmentation fault, reading a pointer that isn't allowed or something.

**Comment by borkdude** [comment]
@jeroenvandijk Now trying this:

```
public static void eval(String expr) {
        System.out.println("x");
        System.out.flush();
        long isolate = createIsolate();
        System.out.println("y");
        System.out.flush();
        CCharPointer c = evalString(isolate);
        System.out.println("z");
        System.out.flush();
        String res = toJavaString(c);
        System.out.println("res: " + res);
        System.out.flush();
        //tearDownIsolate(isolate);
        //return x;
    }
```

It prints:

```
user=> (org.babashka.LibHsqlDb/eval "(+ 1 2 3)")
x
y
debug: xxx
z
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001054fda9a, pid=66525, tid=0x0000000000002103
```

**Comment by retrogradeorbit** [comment]
It's crashing inside `java.lang.StringBuilder::append`

```Stack: [0x00007fb9ec6b1000,0x00007fb9ec7b2000],  sp=0x00007fb9ec7af330,  free space=1016k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
J 681 C2 java.lang.StringBuilder.append(Ljava/lang/String;)Ljava/lang/StringBuilder; (8 bytes) @ 0x00007fb9d52fd6f
5 [0x00007fb9d52fd6c0+0x35]
j  org.babashka.LibHsqlDb.eval(Ljava/lang/String;)Ljava/lang/String;+53
j  user$eval3.invokeStatic()Ljava/lang/Object;+5
j  user$eval3.invoke()Ljava/lang/Object;+0
j  clojure.lang.Compiler.eval(Ljava/lang/Object;Z)Ljava/lang/Object;+374
j  clojure.lang.Compiler.eval(Ljava/lang/Object;)Ljava/lang/Object;+2
j  clojure.core$eval.invokeStatic(Ljava/lang/Object;)Ljava/lang/Object;+3
j  clojure.core$eval.invoke(Ljava/lang/Object;)Ljava/lang/Object;+3
j  clojure.main$repl$read_eval_print__9092$fn__9095.invoke()Ljava/lang/Object;+11
j  clojure.main$repl$read_eval_print__9092.invoke()Ljava/lang/Object;+127
j  clojure.main$repl$fn__9101.invoke()Ljava/lang/Object;+7
j  clojure.main$repl.invokeStatic(Lclojure/lang/ISeq;)Ljava/lang/Object;+639
j  clojure.main$repl_opt.invokeStatic(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;+118
j  clojure.main$main.invokeStatic(Lclojure/lang/ISeq;)Ljava/lang/Object;+462
j  clojure.main$main.doInvoke(Ljava/lang/Object;)Ljava/lang/Object;+6
j  clojure.lang.RestFn.invoke()Ljava/lang/Object;+26
J 467 C1 clojure.lang.AFn.applyToHelper(Lclojure/lang/IFn;Lclojure/lang/ISeq;)Ljava/lang/Object; (3238 bytes) @ 0x00007fb9d52662e4 [0x00007fb9d5250e00+0x154e4]
j  clojure.lang.RestFn.applyTo(Lclojure/lang/ISeq;)Ljava/lang/Object;+23
j  clojure.lang.Var.applyTo(Lclojure/lang/ISeq;)Ljava/lang/Object;+5
j  clojure.main.main([Ljava/lang/String;)V+20
v  ~StubRoutines::call_stub
V  [libjvm.so+0x68825b]  JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*)+0xddb
V  [libjvm.so+0x6d3211]  jni_invoke_static(JNIEnv_*, JavaValue*, _jobject*, JNICallType, _jmethodID*, JNI_ArgumentPusher*, Thread*) [clone .isra.96] [clone .constprop.117]+0x1e1
V  [libjvm.so+0x6d54db]  jni_CallStaticVoidMethod+0x15b
C  [libjli.so+0x889c]  JavaMain+0xa3c
C  [libpthread.so.0+0x76ba]  start_thread+0xca
```

**Comment by borkdude** [comment]
Closing in favor of babashka pods.
